File: Python/ROSDA.py
#Clear Workspace before each run.
import numpy as np
from scipy import linalg

def Psi(z):
    zeta = 0.11  # Establish Tuning Parameter
    Psi = (1 - np.exp(-zeta * z)) / zeta
    return Psi

def Psi_dx(z):
    zeta = 0.11  # Establish tuning parameter
    Psi_dx = np.exp(-zeta * z)
    return Psi_dx

def Theta_0(n, W, X, Py, Q):
    I = np.eye(n)  # Identity matrix
    Pwx = X @ np.linalg.inv(X.T @ W @ X) @ X.T @ W  # Matrix multiplication
    
    # Smallest right singular vectors using SVD
    _, _, V = linalg.svd((I - Pwx) @ W @ Py)
    Theta0 = V[:, -Q:]  # Select the last Q columns
    
    return Theta0

# Establish Matrices and Parameters
tolerance = 1e-14
max_iteration = 10000
RSSold = 2
RSS = 10
iter = 0
K = 2  # Number of classes

# Data
# Synthetic X Matrix
X = np.random.randint(0, 11, size=(100, 40))
X = X / np.linalg.norm(X, axis=0)

# Synthetic Y Matrix - Replace with Target Variable
num_samples = X.shape[0]
num_classes = K  # Classes: 0, 1
Y_labels = np.random.randint(0, 2, size=(num_samples,))

# Convert to dummy variable (one-hot encoding)
Y = np.zeros((num_samples, num_classes))
for i in range(num_samples):
    Y[i, Y_labels[i]] = 1 

n = X.shape[0]
Q = K - 1
betaj = np.ones((X.shape[1], Q))
Thetaj = np.ones((K, Q))
D = (1/n) * (Y.T @ Y)

# ROSDA Loop
while (abs(RSSold - RSS)/RSS > tolerance) and (iter < max_iteration):
    z = np.zeros(n)
    for i in range(n):
        z[i] = np.linalg.norm(Y[i, :] @ Thetaj - X[i, :] @ betaj)**2  # Use default L2 norm
    
    w = np.zeros(n)
    for i in range(n):
        w[i] = Psi_dx(z[i])
    
    W = np.diag(w)
    RSSold = RSS
    
    # betaj calculation
    betaj = np.linalg.inv(X.T @ W @ X) @ X.T @ W @ Y @ Thetaj
    
    # SVD decomposition
    Py, _, _ = linalg.svd(Y)
    I = np.eye(n)
    Theta0 = Theta_0(n, W, X, Py, Q)
    Thetaj = (1/np.sqrt(n)) * np.linalg.inv(D) @ Y.T @ Py @ Theta0
    
    # Recalculate z using theta(j) and beta(j)
    for i in range(n):
        z[i] = np.linalg.norm(Y[i, :] @ Thetaj - X[i, :] @ betaj)**2  # Use default L2 norm
    
    RSS = (1/n) * sum(Psi(z[i]) for i in range(n))
    iter += 1

# Predict classes by finding which Thetak is closest to X_iB
XB = X @ betaj
predicted_classes = np.zeros((n, K))
for i in range(n):
    distances = np.abs(Thetaj - XB[i, :])
    total_distances = np.sum(distances, axis=1)
    index = np.argmin(total_distances)
    predicted_classes[i, index] = 1

# Note: Functions Psi_dx and Theta_0 need to be defined based on the original paper
# References:
# [1] https://numpy.org/doc/stable/reference/generated/numpy.linalg.norm.html
# [2] https://numpy.org/doc/stable/reference/generated/numpy.diag.html
# [3] https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.svd.html
# [4] Python operators follow standard mathematical notation
# [5] For finding closest value: https://numpy.org/doc/stable/reference/generated/numpy.argmin.html

File: MATLAB/ROSDA.m
%OSDA Algorithm Function as presented by Huang & Zhang 2020
%Establish Matrices and Parameters
tolerance = 1e-14;
max_iteration = 10000;
RSSold = 2;
RSS = 10;
iter = 0;
K = 2; %Number of classes

%Data
% Synthetic X Matrix
X = randi([0 10],100, 40); %Replace with Image Matrix
X = normalize(X);

%Synthetic Y Matrix - Replace with Target Variable
num_samples = size(X,1);
num_classes = K; % Classes: 0, 1 
Y_labels = randi([0, 1], num_samples, 1); %Column vector of class labels

% Convert to dummy variable (one-hot encoding)
Y = zeros(num_samples, num_classes); % Initialize matrix
for i = 1:num_samples
    Y(i, Y_labels(i) + 1) = 1; % "+1" because MATLAB indexing starts from 1
end

n = size(X,1);
Q = K-1;
betaj = ones(size(X,2),Q);
Thetaj = ones(K,Q);
D = (1/n)*transpose(Y)*Y;
%ROSDA Loop
while (abs(RSSold - RSS)/RSS > tolerance && iter < max_iteration)
    for i = 1:n
        z(i) = norm(Y(i, :)*Thetaj-X(i, :)*betaj,"fro")^2;%[1]
    end
    for i = 1:n
        w(i) = Psi_dx(z(i));
    end
    W = diag(w);%[2]
    RSSold = RSS;
    betaj = inv(transpose(X)*W*X)*transpose(X)*W*Y*Thetaj;
    [Py,~,~] = svd(Y);
    I = eye(n);
    Theta0 = Theta_0(n,W,X,Py,Q);
    Thetaj = (1/sqrt(n))*inv(D)*transpose(Y)*Py*Theta0;
    for i = 1:n % Recalculate z using theta(j) and beta(j)
        z(i) = norm(Y(i, :) * Thetaj - X(i, :) * betaj, "fro")^2;
    end
    RSS = (1/n) * sum(Psi(z(i)));
    iter=iter+1;
end

% Predict classes by finding which Thetak is closest to X_iB
XB = X * betaj; 
predicted_classes = zeros(n, K);
for i = 1:n
    [~, index] = min(abs(Thetaj - XB(i, :)));
    predicted_classes(i, index) = 1; 
end

%References:
%1. https://www.mathworks.com/help/matlab/ref/norm.html
%2. %https://www.mathworks.com/help/matlab/ref/diag.html
%3. https://www.mathworks.com/help/matlab/ref/double.svd.html#d126e1761848
%4. https://www.mathworks.com/help/matlab/matlab_prog/matlab-operators-and-special-characters.html
% 5. https://www.mathworks.com/matlabcentral/answers/152301-find-closest-value-in-array